README
Port_Scanner_C++
----------------
AUTHOR
	 Allen Bradley Roberts

NAME
	 scan - Scans a target IP address for open ports.

SYNOPSIS
	 ./scan [FLAGS]

FLAG SUMMARY
	 -t <target IP>: Sets target IP. Works with IPv4 only. (Required)
	 -p <target port>: Scans a specific port number. (Optional)

EXAMPLES
	 ./scan -t 127.0.0.1
		 Scans 127.0.0.1 for the 1000 most common ports.

	 ./scan -t 127.0.0.1 -p 80
		 Scans 127.0.0.1 for port 80.


I created a port scanner in python several months ago for fun and figured porting this to C++ would be a good project.
Python was way easier... I was able to find an algorithm on StackOverflow for the logic to check is a port is open. I
was able to improve that and add my own logic to handle the 2 flags and user-input errors.

Running this with a target IP and no port scans the 1000 most common ports according to nmap. This does NOT handle port
ranges such as 1-100. With the -p option with a valid port number only scans that specific port. Having multiple -t or
-p flags does not work, it just exits the program.

CHANGELOG
August 21, 2018 - Initialized CS327 Project

August 25, 2018 - Tested Atom, Vim, Emacs, VSCode, and CLion

August 26, 2018 - Decided on VSCode, though, Vim was a close second. Also set up version control with Git and GitHub.

August 28, 2018 4:00PM - Initialized Knights Tour with psudocode provided by instructor
August 28, 2018 4:30PM - Attempted to use an algorithm I found in a book to help with the knights Tour
August 28, 2018 4:45PM - Gave up on algorithms book version
August 28, 2018 5:30PM - Created a cool algorithm to find all the possible moves a knight can do
August 28, 2018 7:30PM - Trying to figure out how to finish the knights tour algorithm
August 28, 2018 10:00PM - Getting segementation faults and can't figure out why
August 28, 2018 11:00PM - Still getting seg faults and tired.

August 29, 2018 10:30AM - Realized my "cool algorithm" actually sucks, so got rid of that
August 29, 2018 4:30PM - Thinking I have a problem with recursion never ending, so trying to make a base case
August 29, 2018 5:00PM - No longer getting seg faults and printing a single line with 25 numbers, the numbers are wrong
August 29, 2018 6:00PM - Can't figure out anything, looking up algorithms online and found an algorithm that is VERY close so using that to help finish
August 29, 2018 9:00PM - SO, I tried for several hours to get this damn thing to work but nothing seems to help and I don't know what I'm doing wrong
August 29, 2018 10:00PM - Wrote my Makefile, CHANGELOG, and README and calling it. I'm not sure what to do from here
August 29, 2018 9:30PM - Created base directory for all assignments
August 29, 2018 9:45PM - Initialized DungeonGenerator project and added high level

September 2, 2018 9:00PM - Started creating pseudocode for project

September 3. 2018 7:00PM - Starting code of project
September 3, 2018 7:30PM - 80x21 wall created correctly 
September 3, 2018 8:15PM - Created int fetchRand(int N) which creates a random number between 0 and N

September 4, 2018 4:00PM - Worked on figuring out an algorithm for creating the rooms

September 5, 2018 4:00PM - Started working on Room generation
September 5, 2018 6:45PM - Creating random sized and located rooms correctly, now need to add them to the Dungeon
September 5, 2018 7:30PM - Trying to put the dungeon starting x,y position in the dungeon but it's not working correctly
September 5, 2018 7:45PM - Okay, fixed bug, now it's creating the coordinates correctly WITHOUT collision detection. Next step is to print the boxes without colDetection.
September 5, 2018 8:15PM - Rooms are created randomly in the dungeon without collision detection
September 5, 2018 9:00PM - Rooms are goooood, refactored code a bit to make it better
September 5, 2018 9:15PM - Crappy colDetection for rooms made, kinda works, working on the corradors now
September 5, 2018 9:30PM - collision kind of works. It can be super hard on the computer though. it's pretty crappy in all honestly
September 5, 2018 10:00PM - Couldn't get the hallways made, calling it so this assignment isn't late.

September 7, 2018 10:00PM - Grabbed professors dungeon generator code

September 9, 2018 6:00PM - Started reading and learning how professors code works
September 9, 2018 8:00PM - Started save_load.c to learn how to use binary files

September 10, 2018 5:00PM - Writing piece by piece different parts of a binary file by hand and evaluating it in hexdump
September 10, 2018 9:00PM - Have most of the binary file saving properly.

September 11, 2018 7:00PM - Started working on the load functionality
September 12, 2018 3:00AM - Loading now works perfectly with in save_load

September 12, 2018 5:00PM - Started writing code to test all of the professors dungeons in save_load.c
September 12, 2018 5:30PM - Save load is perfect... on to implementing it in professors code.
September 12, 2018 11:00PM - accidently deleted my changelog, going to get the program working then update this.
September 12, 2018 11:00PM - everything works except difference between corrador and halls.
September 12, 2018 11:55PM - Finally works!!!
September 12, 2018 12:15PM - finally figured out big endian stuffz. hope this works perfectly and i get a 10/10

September 20, 2018 12:00PM - Grabbed professors solution for 1.02
September 20, 2018 5:30PM - met with a TA to help get a better grade on 1.02

September 25, 2018 6:00PM - Started assignment 1.03
September 25, 2018 6:30PM - Reformatted professors code because it looks like crappy
September 25, 2018 8:00PM - Can't figure out how Dijkstra's works, so looking at other GitHub solutions
September 25, 2018 11:00PM - Tried to be smart and break up the professors code into .c and .h files and have a main.c run everything... Didnt work at all
September 26, 2018 1:00AM - I understand Dijkstra's better, also wrote some base code to work with that does a basic print of 3 dungeons
September 26, 2018 2:00AM - Copied some code from GitHub to see if it even works and better understand how it works
September 26, 2018 2:30AM - Figured out how it works better, Have tunneling mobs version working, however it has a bug, I think it doesn't display correctly
September 26, 2018 3:30AM - Got non-tunneling map working. Believe it's 100% correct.
September 26, 2018 3:45AM - Pushed everything to GitHub and updating README/CHANGELOG
September 26, 2018 4:00AM - Running with Valgrind to make sure I handle memory correctly.
September 26, 2018 4:00AM part 2 - No memleaks! Also had to install valgrind on ubuntu. 
September 26, 2018 4:00AM part 3 - Actually upon further inspection, I believe this works perfectly.
September 26, 2018 4:00AM part 4 - Turning it in!

October 1, 2018 7:00PM - Got professors code and started by making PC move around.
October 1, 2018 9:00PM - PC moves around randomly

October 2, 2018 6:00PM - Started making Non-tunneling monsters and making them move around.
October 2, 2018 9:00PM - NTMobs move around perfectly.

October 3, 2018 5:00PM - forgot about --nummon. so started that.

test.sh
#!/bin/bash

echo ./scan -t 192.168.1.1
./scan -t 192.168.1.1

echo ./scan -t 192.168.1.1 -p 80
./scan -t 192.168.1.1 -p 80

echo ./scan -t 192.168.1.1 -p 81
./scan -t 192.168.1.1 -p 81

echo ./scan -t 192.168.1.1 -p asdf
./scan -t 192.168.1.1 -p asdf

echo ./scan -t 256.256.256.256
./scan -t 256.256.256.256

echo ./scan -t 256.256.256.256 -p 80
./scan -t 256.256.256.256 -p 80

echo ./scan -t asdf
./scan -t asdf

echo ./scan -t asdf -p 80
./scan -t asdf -p 80

echo ./scan -p 80
./scan -p 80

echo ./scan -p asdf
./scan -p asdf

echo ./scan
./scan

echo ./scan -p 80 -t 192.168.1.1
./scan -p 80 -t 192.168.1.1

echo ./scan -p asdf -t 192.168.1.1
./scan -p asdf -t 192.168.1.1

echo ./scan -password -t 192.168.1.1
./scan -password -t 192.168.1.1

echo ./scan -t 192.168.1.1 -password
./scan -t 192.168.1.1 -password

echo ./scan -t 192.168.1.1 -willThisWork
./scan -t 192.168.1.1 -willThisWork

echo ./scan -t192.158.1.1
./scan -t192.158.1.1

echo ./scan -p 80 -t198.157.3.2
./scan -p 80 -t198.157.3.2

echo ./scan -p 80 -t 192.168.1.1 -p 81
./scan -p 80 -t 192.168.1.1 -p 81

echo ./scan -p 80  -t 192.168.1.1 -t 127.0.0.1
./scan -p 80  -t 192.168.1.1 -t 127.0.0.1

echo "FINISHED TESTS"

src/test.sh
#!/bin/bash

echo ./scan -t 192.168.1.1
./scan -t 192.168.1.1

echo ./scan -t 192.168.1.1 -p 80
./scan -t 192.168.1.1 -p 80

echo ./scan -t 192.168.1.1 -p 81
./scan -t 192.168.1.1 -p 81

echo ./scan -t 192.168.1.1 -p asdf
./scan -t 192.168.1.1 -p asdf

echo ./scan -t 256.256.256.256
./scan -t 256.256.256.256

echo ./scan -t 256.256.256.256 -p 80
./scan -t 256.256.256.256 -p 80

echo ./scan -t asdf
./scan -t asdf

echo ./scan -t asdf -p 80
./scan -t asdf -p 80

echo ./scan -p 80
./scan -p 80

echo ./scan -p asdf
./scan -p asdf

echo ./scan
./scan

echo ./scan -p 80 -t 192.168.1.1
./scan -p 80 -t 192.168.1.1

echo ./scan -p asdf -t 192.168.1.1
./scan -p asdf -t 192.168.1.1

echo ./scan -password -t 192.168.1.1
./scan -password -t 192.168.1.1

echo ./scan -t 192.168.1.1 -password
./scan -t 192.168.1.1 -password

echo ./scan -t 192.168.1.1 -willThisWork
./scan -t 192.168.1.1 -willThisWork

echo ./scan -t192.158.1.1
./scan -t192.158.1.1

echo ./scan -p 80 -t198.157.3.2
./scan -p 80 -t198.157.3.2

echo ./scan -p 80 -t 192.168.1.1 -p 81
./scan -p 80 -t 192.168.1.1 -p 81

echo ./scan -p 80  -t 192.168.1.1 -t 127.0.0.1
./scan -p 80  -t 192.168.1.1 -t 127.0.0.1

echo ./scan -t 192.168.1.100
./scan -t 192.168.1.100

echo "FINISHED TESTS"

scanner.cpp
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <iostream>
#include <string>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <netdb.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

static bool port_open(std::string ip, int port);


int main(int argc, char *argv[]){
    
    std::string target_ip = "10.142.196.238";
    int target_port;
    bool scanned = false;

    for(int i = 1; i <= 1000; i++){
        target_port = i;
        if(port_open(target_ip, target_port)){
            scanned = true;
            std::cout << target_ip << ":" << target_port << " is open." << std::endl;
        } 
    }

    if(!scanned){
        std::cout << "Nothing scanned." << std::endl;
    }

    return 0;
}


// Found this on StackOverflow
static bool port_open(std::string ip, int port){
    short int sock = -1;
    fd_set fd;
    struct sockaddr_in address;
    struct timeval tv;

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = inet_addr(ip.c_str());
    address.sin_port = htons(port);

    sock = socket(AF_INET, SOCK_STREAM, 0);
    fcntl(sock, F_SETFL, O_NONBLOCK);

    connect(sock, (struct sockaddr *)&address, sizeof(address));

    FD_ZERO(&fd);
    FD_SET(sock, &fd);
    tv.tv_sec = 0;
    tv.tv_usec = 30000; // Increased this time so it actually does something.

    if(select(sock+1, NULL, &fd, NULL, &tv) == 1){
        int error;
        socklen_t len = sizeof(int);
        getsockopt(sock, SOL_SOCKET, SO_ERROR, &error, &len);

        if(error == 0){
            close(sock);
            return true;
        } else {
            close(sock);
            return false;
        }
    }
    return false;
}


src/main.cpp
#include <arpa/inet.h>
#include <iostream>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>

static bool port_open(std::string ip, uint16_t port);
static int host_up(std::string ip);
static void run_scan(std::string ip, uint16_t port);


struct sockaddr_in address;
struct timeval tv;

uint16_t common_ports[] = {
                      1,3,4,6,7,9,13,17,19,20,21,22,23,24,25,26,30,32,33,37,42,43,49,53,70,79,80,81,82,83,84,85,88,89,
                      90,99,100,106,109,110,111,113,119,125,135,139,143,144,146,161,163,179,199,211,212,222,254,255,256,
                      259,264,280,301,306,311,340,366,389,406,407,416,417,425,427,443,444,445,458,464,465,481,497,500,
                      512,513,514,515,524,541,543,544,545,548,554,555,563,587,593,616,617,625,631,636,646,648,666,667,
                      668,683,687,691,700,705,711,714,720,722,726,749,765,777,783,787,800,801,808,843,873,880,888,898,
                      900,901,902,903,911,912,981,987,990,992,993,995,999,1000,1001,1002,1007,1009,1010,1011,1021,1022,
                      1023,1024,1025,1026,1027,1028,1029,1030,1031,1032,1033,1034,1035,1036,1037,1038,1039,1040,1041,
                      1042,1043,1044,1045,1046,1047,1048,1049,1050,1051,1052,1053,1054,1055,1056,1057,1058,1059,1060,
                      1061,1062,1063,1064,1065,1066,1067,1068,1069,1070,1071,1072,1073,1074,1075,1076,1077,1078,1079,
                      1080,1081,1082,1083,1084,1085,1086,1087,1088,1089,1090,1091,1092,1093,1094,1095,1096,1097,1098,
                      1099,1100,1102,1104,1105,1106,1107,1108,1110,1111,1112,1113,1114,1117,1119,1121,1122,1123,1124,
                      1126,1130,1131,1132,1137,1138,1141,1145,1147,1148,1149,1151,1152,1154,1163,1164,1165,1166,1169,
                      1174,1175,1183,1185,1186,1187,1192,1198,1199,1201,1213,1216,1217,1218,1233,1234,1236,1244,1247,
                      1248,1259,1271,1272,1277,1287,1296,1300,1301,1309,1310,1311,1322,1328,1334,1352,1417,1433,1434,
                      1443,1455,1461,1494,1500,1501,1503,1521,1524,1533,1556,1580,1583,1594,1600,1641,1658,1666,1687,
                      1688,1700,1717,1718,1719,1720,1721,1723,1755,1761,1782,1783,1801,1805,1812,1839,1840,1862,1863,
                      1864,1875,1900,1914,1935,1947,1971,1972,1974,1984,1998,1999,2000,2001,2002,2003,2004,2005,2006,
                      2007,2008,2009,2010,2013,2020,2021,2022,2030,2033,2034,2035,2038,2040,2041,2042,2043,2045,2046,
                      2047,2048,2049,2065,2068,2099,2100,2103,2105,2106,2107,2111,2119,2121,2126,2135,2144,2160,2161,
                      2170,2179,2190,2191,2196,2200,2222,2251,2260,2288,2301,2323,2366,2381,2382,2383,2393,2394,2399,
                      2401,2492,2500,2522,2525,2557,2601,2602,2604,2605,2607,2608,2638,2701,2702,2710,2717,2718,2725,
                      2800,2809,2811,2869,2875,2909,2910,2920,2967,2968,2998,3000,3001,3003,3005,3006,3007,3011,3013,
                      3017,3030,3031,3052,3071,3077,3128,3168,3211,3221,3260,3261,3268,3269,3283,3300,3301,3306,3322,
                      3323,3324,3325,3333,3351,3367,3369,3370,3371,3372,3389,3390,3404,3476,3493,3517,3527,3546,3551,
                      3580,3659,3689,3690,3703,3737,3766,3784,3800,3801,3809,3814,3826,3827,3828,3851,3869,3871,3878,
                      3880,3889,3905,3914,3918,3920,3945,3971,3986,3995,3998,4000,4001,4002,4003,4004,4005,4006,4045,
                      4111,4125,4126,4129,4224,4242,4279,4321,4343,4443,4444,4445,4446,4449,4550,4567,4662,4848,4899,
                      4900,4998,5000,5001,5002,5003,5004,5009,5030,5033,5050,5051,5054,5060,5061,5080,5087,5100,5101,
                      5102,5120,5190,5200,5214,5221,5222,5225,5226,5269,5280,5298,5357,5405,5414,5431,5432,5440,5500,
                      5510,5544,5550,5555,5560,5566,5631,5633,5666,5678,5679,5718,5730,5800,5801,5802,5810,5811,5815,
                      5822,5825,5850,5859,5862,5877,5900,5901,5902,5903,5904,5906,5907,5910,5911,5915,5922,5925,5950,
                      5952,5959,5960,5961,5962,5963,5987,5988,5989,5998,5999,6000,6001,6002,6003,6004,6005,6006,6007,
                      6009,6025,6059,6100,6101,6106,6112,6123,6129,6156,6346,6389,6502,6510,6543,6547,6565,6566,6567,
                      6580,6646,6666,6667,6668,6669,6689,6692,6699,6779,6788,6789,6792,6839,6881,6901,6969,7000,7001,
                      7002,7004,7007,7019,7025,7070,7100,7103,7106,7200,7201,7402,7435,7443,7496,7512,7625,7627,7676,
                      7741,7777,7778,7800,7911,7920,7921,7937,7938,7999,8000,8001,8002,8007,8008,8009,8010,8011,8021,
                      8022,8031,8042,8045,8080,8081,8082,8083,8084,8085,8086,8087,8088,8089,8090,8093,8099,8100,8180,
                      8181,8192,8193,8194,8200,8222,8254,8290,8291,8292,8300,8333,8383,8400,8402,8443,8500,8600,8649,
                      8651,8652,8654,8701,8800,8873,8888,8899,8994,9000,9001,9002,9003,9009,9010,9011,9040,9050,9071,
                      9080,9081,9090,9091,9099,9100,9101,9102,9103,9110,9111,9200,9207,9220,9290,9415,9418,9485,9500,
                      9502,9503,9535,9575,9593,9594,9595,9618,9666,9876,9877,9878,9898,9900,9917,9929,9943,9944,9968,
                      9998,9999,10000,10001,10002,10003,10004,10009,10010,10012,10024,10025,10082,10180,10215,10243,
                      10566,10616,10617,10621,10626,10628,10629,10778,11110,11111,11967,12000,12174,12265,12345,13456,
                      13722,13782,13783,14000,14238,14441,14442,15000,15002,15003,15004,15660,15742,16000,16001,16012,
                      16016,16018,16080,16113,16992,16993,17877,17988,18040,18101,18988,19101,19283,19315,19350,19780,
                      19801,19842,20000,20005,20031,20221,20222,20828,21571,22939,23502,24444,24800,25734,25735,26214,
                      27000,27352,27353,27355,27356,27715,28201,30000,30718,30951,31038,31337,32768,32769,32770,32771,
                      32772,32773,32774,32775,32776,32777,32778,32779,32780,32781,32782,32783,32784,32785,33354,33899,
                      34571,34572,34573,35500,38292,40193,40911,41511,42510,44176,44442,44443,44501,45100,48080,49152,
                      49153,49154,49155,49156,49157,49158,49159,49160,49161,49163,49165,49167,49175,49176,49400,49999,
                      50000,50001,50002,50003,50006,50300,50389,50500,50636,50800,51103,51493,52673,52822,52848,52869,
                      54045,54328,55055,55056,55555,55600,56737,56738,57294,57797,58080,60020,60443,61532,61900,62078,
                      63331,64623,64680,65000,65129,65389
};

std::string usage = "AUTHOR\n"
                    "\t Allen Bradley Roberts\n\n"
                    "NAME\n"
                    "\t scan - Scans a target IP address for open ports.\n\n"
                    "SYNOPSIS\n"
                    "\t ./scan [FLAGS]\n\n"
                    "FLAG SUMMARY\n"
                    "\t -t <target IP>: Sets target IP. Works with IPv4 only. (Required) \n"
                    "\t -p <target port>: Scans a specific port number. (Optional)\n\n"
                    "EXAMPLES\n"
                    "\t ./scan -t 127.0.0.1\n"
                    "\t\t Scans 127.0.0.1 for the 1000 most common ports.\n\n"
                    "\t ./scan -t 127.0.0.1 -p 80\n"
                    "\t\t Scans 127.0.0.1 for port 80.\n\n";

int main(int argc, char *argv[]){
    std::string target_ip;
    uint16_t i;
    int port = 0;
    bool target_set = false;
    struct sockaddr_in sa;
    bool t = false;
    bool p = false;

   if (argc > 1) {
        for (i = 1; i < argc; i++) {
            if (argv[i][0] == '-') {
                switch (argv[i][1]) {
                    case 'p':
                        if(argv[i][2]){
                            std::cout << "Invalid input. A space is required between -p and the port number."
                                      << std::endl
                                      << "Run ./scan without any arguments to display usage." << std::endl;
                            return -1;
                        }else if (p) {
                            std::cout << "Error - Invalid arguemts. Please only use one port flag." << std::endl;
                            return -1;
                        }else {
                                port = atoi(argv[i + 1]);

                                if (port < 1 || port > 65535) {
                                    std::cout << "Invalid port number. Please enter an integer between 1 - 65535."
                                    << std::endl
                                    << "Run ./scan without any arguments to display usage." << std::endl;
                                    port = -1;
                                } else {
                                    port = (uint16_t) atoi(argv[i + 1]);
                                }
                        }

                        p = true;
                        break;

                    case 't':
                        if(argv[i][2]){
                            std::cout << "Invalid input. A space is required between -t and the port number."
                                      << std::endl
                                      << "Run ./scan without any arguments to display usage." << std::endl;
                            return -1;
                        }else if(t) {
                            std::cout << "Error - Invalid arguments. Please only use one target flag."
                                      << std::endl;
                            return -1;
                        }else {
                                target_ip = argv[i + 1];
                                if (inet_pton(AF_INET, target_ip.c_str(), &sa.sin_addr)) {
                                    target_set = true;
                                }
                            }

                        t = true;
                        break;

                    default:
                        std::cout << "Invalid Arguments." << std::endl << "Run ./scan without any arguments to display"
                                                                          "usage." << std::endl;
                        return -1;
                }
            }
        }
   } else {
       std::cout << usage << std::endl;
   }


   if(target_set) {
       run_scan(target_ip, port);
   }

   if(!target_set && argc > 1){
       std::cout << "Error - Invalid IP address. Please enter a valid IPv4 address." <<
       std::endl << "Run ./scan without any arguments to display usage." << std::endl;
   }

   std::cout << std::endl <<"Finished." << std::endl;

   return 0;
}


static void run_scan(std::string ip, uint16_t port){
    bool scanned = false;
    int count = 0;

    //int ping = system(("ping -c 1 " + target_ip + "| grep -q \"0 received\"").c_str());
    // I thought this was super clever but might have not worked depending on OS... not sure...

    if (host_up(ip) == 1) {
        if (port == 0) {
            std::cout << "Scanning " << ip << " for 1000 most common ports." << std::endl << std::endl;
            for (uint16_t ports : common_ports) {
                if (port_open(ip, ports)) {
                    scanned = true;
                    count++;
                    std::cout << "Port: " << ports << " is open." << std::endl;
                }
            }
        } else if (port > 0) {
            std::cout << "Scanning " << ip << " port " << port << "." << std::endl << std::endl;
            if (port_open(ip, port)) {
                scanned = true;
                count++;
                std::cout << "Port: " << port << " is open." << std::endl;
            } else {
                std::cout << "Port: " << port << " is closed." << std::endl;
            }
        }
    } else {
        std::cout << "Unable to connect to target. Target is either down or blocking ping probes." << std::endl;
    }

    if (scanned) {
        std::cout << std::endl<< "Number of opened ports: " << count << std::endl;
    }
}

static bool port_open(std::string ip, uint16_t port){
    int sock;
    int error = -1;
    fd_set fd;

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = inet_addr(ip.c_str());
    address.sin_port = htons(port);

    sock = socket(AF_INET, SOCK_STREAM, 0);
    fcntl(sock, F_SETFL, O_NONBLOCK);

    connect(sock, (struct sockaddr *)&address, sizeof(address));

    FD_ZERO(&fd);
    FD_SET(sock, &fd);
    tv.tv_sec = 2;

    if(select(sock+1, nullptr , &fd, nullptr, &tv) == 1){
        socklen_t len = sizeof(int);
        getsockopt(sock, SOL_SOCKET, SO_ERROR, &error, &len);

        close(sock);
    }

    return !error;
}


static int host_up(std::string ip){
    int sock;
    fd_set fd;

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = inet_addr(ip.c_str());

    sock = socket(AF_INET, SOCK_STREAM, 0);
    fcntl(sock, F_SETFL, O_NONBLOCK);

    connect(sock, (struct sockaddr *)&address, sizeof(address));

    FD_ZERO(&fd);
    FD_SET(sock, &fd);
    tv.tv_sec = 2;

    return select(sock+1, nullptr , &fd, nullptr, &tv);
}